File: domain/src/main/java/AggregateRootBase.java
import java.util.ArrayList;
import java.util.List;

public abstract class AggregateRootBase<TMemento extends Memento> implements Aggregate {
    private final String _id;
    private final List<DomainEvent> _uncommittedEvents = new ArrayList<>();
    private final IRouteEvents _registeredRoutes = null;
    private Integer _version;

    protected AggregateRootBase(String id, Memento memento) {
        _id = id;
        internalApplySnapshot(memento);
    }

    public String getId() {
        return _id;
    }

    public int getVersion() {
        return _version;
    }

    public List<DomainEvent> getUncommittedEvents() {
        return _uncommittedEvents;
    }

    public Memento getSnapshot() {
        return createSnapshot(_id, _version);
    }

    protected abstract <TMemento extends Memento> void applySnapshot(TMemento memento);

    protected abstract Memento createSnapshot(String id, Integer version);

    private void internalApplySnapshot(Memento snapshot) {
        if (snapshot != null) {
            _version = snapshot.getVersion();
            applySnapshot(snapshot);
        }
    }

    @Override
    public void applyEvent(Object evt) {
        _registeredRoutes.dispatch(evt);
        _version++;
    }

    @Override
    public void clearUncommittedEvents() {
        _uncommittedEvents.clear();
    }
}

File: modules/domain/src/main/kotlin/org/openmedstack/domain/AggregateRootBase.kt
package org.openmedstack.domain

import org.openmedstack.events.DomainEvent

abstract class AggregateRootBase<TMemento : Memento> protected constructor(id: String, memento: TMemento?) : Aggregate {
    private val _uncommittedEvents: MutableList<DomainEvent> = ArrayList()
    private val _registeredRoutes: IRouteEvents
    private val _id: String
    private var _version: Int = 0

    override val id: String
        get() = _id

    override val version: Int
        get() = _version

    override val uncommittedEvents: MutableList<DomainEvent>
        get() = _uncommittedEvents

    override fun getSnapshot(): Memento {
        return createSnapshot(_id, _version)
    }

    protected abstract fun applySnapshot(memento: TMemento)

    protected abstract fun createSnapshot(id: String?, version: Int?): Memento

    private fun internalApplySnapshot(snapshot: TMemento?) {
        if (snapshot != null) {
            _version = snapshot.version
            applySnapshot(snapshot)
        }
    }

    @Throws(HandlerForDomainEventNotFoundException::class)
    override fun applyEvent(event: DomainEvent) {
        _registeredRoutes.dispatch(event, event.javaClass)
        _version++
    }

    override fun clearUncommittedEvents() {
        _uncommittedEvents.clear()
    }

    init {
        _id = id
        _registeredRoutes = ConventionEventRouter(true, this)
        internalApplySnapshot(memento)
    }
}